source.go

```diff
@@ -1,60 +1,145 @@
-package connectorname
+// Copyright © 2023 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package weather
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
 
 	sdk "github.com/conduitio/conduit-connector-sdk"
+	"golang.org/x/time/rate"
 )
 
 type Source struct {
 	sdk.UnimplementedSource
 
-	config           SourceConfig
-	lastPositionRead sdk.Position //nolint:unused // this is just an example
+	config  SourceConfig
+	client  *http.Client
+	url     string
+	limiter *rate.Limiter
 }
 
 func NewSource() sdk.Source {
 	return &Source{}
 }
 
 func (s *Source) Parameters() map[string]sdk.Parameter {
-	return map[string]sdk.Parameter{
-		GlobalConfigParam: {
-			Default:     "localhost:10000",
-			Required:    true,
-			Description: "The URL of the server.",
-		},
-		SourceConfigParam: {
-			Default:     "false",
-			Required:    false,
-			Description: "An optional configuration parameter.",
-		},
-	}
+	return s.config.Parameters()
 }
 
 func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
-	sdk.Logger(ctx).Info().Msg("Configuring a Source Connector...")
-	config, err := ParseSourceConfig(cfg)
+	sdk.Logger(ctx).Info().Msg("Configuring Weather Source Connector...")
+	var config SourceConfig
+	err := sdk.Util.ParseConfig(cfg, &config)
 	if err != nil {
 		return err
 	}
-
 	s.config = config
 	return nil
 }
 
 func (s *Source) Open(ctx context.Context, pos sdk.Position) error {
+	s.client = &http.Client{}
+	s.url = s.CreateRequestURL()
+	// try pinging the URL with APPID
+	req, err := http.NewRequestWithContext(ctx, http.MethodHead, s.url, nil)
+	if err != nil {
+		return fmt.Errorf("error creating HTTP request %q: %w", s.config.URL, err)
+	}
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("error pinging URL %q: %w", s.config.URL, err)
+	}
+	if resp.StatusCode == http.StatusUnauthorized {
+		return fmt.Errorf("authorization failed, check your APPID key")
+	}
+	defer resp.Body.Close()
+
+	s.limiter = rate.NewLimiter(rate.Every(s.config.PollingPeriod), 1)
+
 	return nil
 }
 
 func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
-	return sdk.Record{}, nil
+	if s.limiter.Allow() {
```

Comment:
It would be nicer to use `s.limiter.Wait(ctx)`, `Read` is allowed to block until either the context is cancelled or there is a record that can be returned. Then we don't have to "abuse" the SDK and do backoff retry.

Reply:
oh I didn't go with `wait` at first cuz i didn't want the connector to block for 5 minutes, but adding the context to be checked looks better!

Reply:
done